### PR TITLE
ENG2STAAR-2224: Fix bug in `clangmetatool::SourceUtil::isPartialMacro`

### PR DESCRIPTION
**Pull request by envp**
*Issue number of the reported bug or feature request: #<number>*
ENG2STAAR-2224

**Describe your changes**
Replacements created for source ranges overlapping with token pasted
`__VA_ARGS__` are currently faulty. This is a quick fix which can be used
by downstream systems to issue warnings.

**Testing performed**
New test data has been added as the case`033-detect-partial-macro-expansion`. This includes a test driver for the positve & negative cases and accompanying source files with the syntax under test.

### CODE REVIEW

**Comment by envp** [comment]
cc: @dbeer1 @HaltCatchFire @begroh

File: src/source_util.cpp [hunk 1]
```diff
@@ -220,6 +231,7 @@ bool SourceUtil::isPartialMacro(const clang::SourceRange &sourceRange,
     // Only process macros where the statement is in the body, not ones where
     // it is an argument.
 
+    printMacroTokens(begin, sourceManager, preprocessor);
```
Comment (dbeer1) [inline src/source_util.cpp:234]: Remove debug code

File: src/source_util.cpp [hunk 2]
```diff
@@ -257,13 +269,33 @@ bool SourceUtil::isPartialMacro(const clang::SourceRange &sourceRange,
...
+
+      // We don't play well with variadic token pasting inside of macros
+      // currently return true
+      if (currentMacro->isVariadic()) {
+        auto hashhashIt = std::find_if(currentMacroDefTokens.begin(),
+                                       currentMacroDefTokens.end(),
+                                       [](const clang::Token &tok) {
+                                         return tok.is(clang::tok::hashhash);
+                                       });
+        // If there is token pasting, ensure that its not followed by variadic
```
Comment (dbeer1) [inline src/source_util.cpp:289]: It's not token pasting, it's a gcc extension to __VA_ARGS__ :)

File: src/source_util.cpp [hunk 3]
```diff
@@ -257,13 +269,33 @@ bool SourceUtil::isPartialMacro(const clang::SourceRange &sourceRange,
...
+        // args
+        if (hashhashIt != currentMacroDefTokens.end()) {
```
Comment (dbeer1) [inline src/source_util.cpp:291]: We should probably check this in a loop, right? We could have multiple instances of double hashes, and the one before a `__VA_ARGS__` might not be the first one.
```
#define test(a, ...) f(a ## .value(), ##__VA_ARGS__)
```

File: t/033-detect-partial-macro-expansion.t.cpp [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  clang::tooling::RefactoringTool tool(optionsParser.getCompilations(),
+                                       optionsParser.getSourcePathList());
+  bool expectedResult = true;
+  clangmetatool::MetaToolFactory<clangmetatool::MetaTool<TestTool>> raf(
+      tool.getReplacements(), expectedResult);
+  int r = tool.runAndSave(&raf);
+  ASSERT_EQ(0, r);
+}
+
+TEST(partialMacroExpansion_Detect, negative) {
```
Comment (dbeer1) [inline t/033-detect-partial-macro-expansion.t.cpp:105]: Did you want to test the positive ones too? A more straightforward way might be put both test cases in the same .cpp file, and find all variable accesses that have ancestor of a specific function name, eg name the functions `positive` and `negative`, then add `hasAncestor(functionDecl(hasName(x)))` to your matcher and pass in `positive` and `negative` when constructing the test cases

**Comment by envp** [comment]
@dbeer1 Sorry, forgot to delete an unused header.

**Comment by envp** [comment]
cc: @azeemba @ruoso for merge

**Pull request merged**